subsys/bluetooth/host/iso.c

```diff
@@ -1096,3 +1105,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+			bt_iso_chan_set_state(big->bis[i], BT_ISO_CONNECTED);
+		}
+
+		bigs[big->handle] = big;
+	}
+
+	net_buf_unref(rsp);
+
+	return err;
+
+}
+
+int bt_iso_create_big(struct bt_le_ext_adv *padv,
+		      struct bt_iso_create_big_param_t *param)
+{
+	struct bt_iso_big *big = param->big;
+	int err;
+
+	if (!atomic_test_bit(padv->flags, BT_PER_ADV_PARAMS_SET)) {
+		BT_DBG("PA params not set; invalid adv object");
+		return -EINVAL;
+	}
+
+	if (big->handle) {
+		BT_DBG("BIG handle already initialized %u", big->handle);
+		return -EALREADY;
+	}
+
+	err = big_init_bis(big, true);
+	if (err) {
+		BT_DBG("Could not init BIG %d", err);
+		cleanup_big(big);
+		return err;
+	}
+
+	err = hci_le_create_big(padv, param);
+	if (err) {
+		BT_DBG("Could not create BIG %d", err);
+		cleanup_big(big);
+		return err;
+	}
+
+	return err;
+}
+
+static int hci_le_terminate_big(struct bt_iso_big *big)
+{
+	struct bt_hci_cp_le_terminate_big *req;
+	struct bt_hci_evt_le_big_terminate *evt;
+	struct net_buf *buf;
+	struct net_buf *rsp;
+	int err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_TERMINATE_BIG, sizeof(*req));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	req = net_buf_add(buf, sizeof(*req));
+	req->big_handle = big->handle;
+	req->reason = BT_HCI_ERR_REMOTE_USER_TERM_CONN;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_TERMINATE_BIG, buf, &rsp);
+	if (err) {
+		return err;
+	}
+
+	evt = (struct bt_hci_evt_le_big_terminate *)rsp->data;
+	if (evt->big_handle != big->handle) {
+		err = -EIO;
+	} else {
+		big_disconnect(big);
+	}
+
+	return err;
+
+}
+
+static int hci_le_big_sync_term(struct bt_iso_big *big)
+{
+	struct bt_hci_cp_le_big_terminate_sync *req;
+	struct bt_hci_rp_le_big_terminate_sync *evt;
+	struct net_buf *buf;
+	struct net_buf *rsp;
+	int err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_BIG_TERMINATE_SYNC, sizeof(*req));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	req = net_buf_add(buf, sizeof(*req));
+	req->big_handle = big->handle;
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_BIG_TERMINATE_SYNC, buf, &rsp);
+	if (err) {
+		return err;
+	}
+
+	evt = (struct bt_hci_rp_le_big_terminate_sync *)rsp->data;
+	if (evt->status || (evt->big_handle != big->handle)) {
+		err = -EIO;
+	} else {
+		big_disconnect(big);
+	}
+
+	net_buf_unref(rsp);
+
+	return err;
+}
+
+int bt_iso_terminate_big(struct bt_iso_big *big)
+{
+	int err;
+	bool broadcaster;
+
+	if (!big->handle || !big->num_bis || !big->bis) {
+		BT_DBG("BIG not initialized");
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < big->num_bis; i++) {
+		if (!big->bis[i]) {
+			BT_DBG("BIG BIS[%d] not initialized", i);
+			return -EINVAL;
+		}
+	}
+
+	/* They all have the same QOS dir so we can just check the first */
+	broadcaster = big->bis[0]->qos->dir == BT_ISO_CHAN_QOS_OUT;
+
+	if (broadcaster) {
+		err = hci_le_terminate_big(big);
+	} else {
+		err = hci_le_big_sync_term(big);
+	}
+
+	if (err) {
+		BT_DBG("Could not terminate BIG %d", err);
+	} else {
+		bigs[big->handle] = NULL;
+		cleanup_big(big);
+	}
```

Comment:
Missing the big picture: What is done if there is an error terminating?

Reply:
For BIGs this actually cannot fail (only internal host states are updated, and all functions are even `void`).

Reply:
I was thinking about the error returned from the hci commands.

Reply:
There are no HCI commands for cleanup_big :) All cleanup is done on the host side only. `cleanup_big` is more often than not, used to actually handle host cleanup if any HCI commands fail.